src/Concordium/Client/GRPC2.hs

```diff
@@ -2533,6 +2536,303 @@ instance FromProto Proto.PendingUpdate where
+                 { _grpcClientConfigCompression = uncompressed
+                 , _grpcClientConfigHeaders = header
+                 , _grpcClientConfigTimeout = Timeout (fromMaybe 300 (timeout config))
+                 }
+   in liftIO $ do
+       lock <- initializeLock
+       ioref <- newIORef Nothing -- don't start the connection just now
+       killConnection <- newEmptyMVar
+       let logger = fromMaybe (const (return ())) mLogger
+       return $! EnvData (retryNum config) cfg lock ioref logger killConnection
+
+-- |Cookie headers that may be returned by the node in a query.
+type CookieHeaders = Map.Map BS8.ByteString BS8.ByteString
+
+
+-- |A reader-writer lock that strongly prefer writers. More precisely this means the following
+-- - readers and writers are mutually exclusive
+-- - multiple readers may hold the lock at the same time if there is no writer
+-- - at most one writer may hold the lock
+--
+-- If a writer tries to acquire a lock it will either
+-- - succeed if there are no current readers or writers
+-- - block after recording the intent to lock. While there are pending writers no new readers can acquire the lock.
+--
+-- If multiple writers are blocking on the lock they will be served in an
+-- unspecified order and in principle it is possible that with heavy write
+-- contention some writers would be starved. This is not the case for the
+-- use-case we have.
+--
+-- Let ⊤ mean that the MVar is full and ⊥ that it is empty. The fields of the lock satisfy the following
+-- properties.
+-- - there are exactly waitingWriters threads blocking on acquireWrite
+-- - rwlState == Free if and only if rwlReadLock == ⊤ and rwlWriteLock == ⊤
+-- - rwlReadLock == ⊥ if and only if there is an active reader.
+-- - rwlWriteLock == ⊥ if and only if there is an active writer.
+--
+-- Transitions between states are governed by the following transition system
+-- where AW/RW and AR/RR mean acquire write, release write and acquire read,
+-- release read, respectively. The WR and WW mean that the thread that
+-- executed the transition is blocked waiting for rwlReadLock and rwlWriteLock MVar to be full.
+-- (Free 0, ⊤, ⊤) -AR-> (ReadLocked 1 0, ⊥, ⊤)
+-- (Free (n+1), ⊤, ⊤) -AR-> (Free (n+1), ⊤, ⊤)
+-- (Free 0, ⊤, ⊤) -AW-> (WriteLocked 0, ⊤, ⊥)
+-- (Free (n+1), ⊤, ⊤) -AW-> (WriteLocked n, ⊤, ⊥)
+--
+-- (ReadLocked n 0, ⊥, ⊤) -AR-> (ReadLocked (n+1) 0, ⊥, ⊤)
+-- (ReadLocked n (m+1), ⊥, ⊤) -AR-> (ReadLocked n (m+1), ⊥, ⊤), WR
+-- (ReadLocked n m, ⊥, ⊤) -AW-> (ReadLocked n (m+1), ⊥, ⊤), WW
+-- (ReadLocked 1 m, ⊥, ⊤) -RR-> (Free m, ⊤, ⊤)
+-- (ReadLocked (n+1) m, ⊥, ⊤) -RR-> (ReadLocked n m, ⊥, ⊤)
+--
+-- (WriteLocked n, ⊤, ⊥) -AR-> (WriteLocked n, ⊤, ⊥)
+-- (WriteLocked n, ⊤, ⊥) -AW-> (WriteLocked (n+1), ⊤, ⊥), WR
+-- (WriteLocked n, ⊤, ⊥) -RW-> (Free n, ⊤, ⊤), WW
+--
+-- No other state should be reachable.
+--
+-- Additionally, rwlReadLock and rwlWriteLock can only be modified while the
+-- rwlState MVar is held.
+data RWLock = RWLock {
```

Comment:
Since you are moving this lock, I think you should put it in its own file.

Reply:
Do you want a file which contains this lock, or are you referring to my comment here:

https://github.com/Concordium/concordium-client/pull/214#issuecomment-1433282886

As mentioned I think it makes sense to split the current GRPC2 module :-)

Reply:
If you move the lock into its own module **in its own PR** then I don't think there is much advantage in splitting the module further.

I don't mind it either, but it needs to be done in self-contained pull requests that are simple to review.

Reply:
Resolved due to the discussion in another thread.